### PR TITLE
Channel page: rearchitect rows, tabs; add bots and attachments

### PR DESCRIPTION
**Pull request by jakob223**
cc @keybase/y2ksquad

### CODE REVIEW

**Comment by jakob223** [comment]
image
image

File: shared/teams/channel/index.tsx [hunk 1]
```diff
@@ -4,67 +4,210 @@ import * as Types from '../../constants/types/teams'
...
+
+  React.useEffect(() => {
+    if (conversationIDKey !== prevConvID) {
+      setSelectedTab(defaultSelectedTab)
+    }
+  }, [conversationIDKey, prevConvID, setSelectedTab, defaultSelectedTab])
+  return [selectedTab, setSelectedTab]
+}
+
+const emptyMap = new Map()
```
Comment (jzila) [inline shared/teams/channel/index.tsx]: Why is this declared out here? If we use it twice, the two uses would reference the same map.

File: shared/chat/conversation/info-panel/attachments.tsx [hunk 2]
```diff
@@ -628,7 +636,10 @@ export default (p: Props) => {
         break
     }
   }
-
+  return sections
+}
```
Comment (jzila) [inline shared/chat/conversation/info-panel/attachments.tsx]: extra line of whitespace pls

File: shared/common-adapters/section-list.d.ts [hunk 3]
```diff
@@ -7,6 +7,20 @@ type DesktopProps = {
...
+export interface SectionBase<ItemT> {
+  data: ReadonlyArray<ItemT>
+
+  key?: string
+
+  renderItem?: ({index: number, item: ItemT}) => React.ReactNode | null
+
+  ItemSeparatorComponent?: React.ComponentType<any> | null
+
+  keyExtractor?: (item: ItemT, index: number) => string
```
Comment (jzila) [inline shared/common-adapters/section-list.d.ts]: Why so much whitespace between the lines?

Comment (jzila) [reply shared/common-adapters/section-list.d.ts]: And where are `ItemSeparatorComponent` and `keyExtractor` used? I can't find them anywhere in this PR.

File: shared/teams/channel/index.tsx [hunk 4]
```diff
@@ -4,67 +4,210 @@ import * as Types from '../../constants/types/teams'
...
+const Channel = (props: OwnProps) => {
+  const teamID = Container.getRouteProps(props, 'teamID', Types.noTeamID)
+  const conversationIDKey = Container.getRouteProps(props, 'conversationIDKey', '')
+  const providedTab = Container.getRouteProps(props, 'selectedTab', undefined)
+
+  const {bots, participants} = Container.useSelector(state =>
+    ChatConstants.getBotsAndParticipants(state, conversationIDKey)
```
Comment (jzila) [inline shared/teams/channel/index.tsx]: `getBotsAndParticipants(state, conversationIDKey, true /* sort */)`. You need this so the participants are consistently sorted. Also, the bots aren't currently sorted in that function so you should add that there to respect the `sort` variable.

**Comment by jzila** [comment]
BTW in the future please separate refactors into a different PR from new code. Cleaner for reviewing and also cleaner for merge history in master.

File: shared/teams/channel/index.tsx (hunk 1 above)
Comment (jakob223) [reply shared/teams/channel/index.tsx]: Exactly. This way, when `Container.useSelector(state => state.teams.teamIDToMembers.get(teamID) ?? emptyMap)` returns it twice, they aren't mistaken for different things, triggering a pointless rerender.

File: shared/common-adapters/section-list.d.ts (hunk 3 above)
Comment (jakob223) [reply shared/common-adapters/section-list.d.ts]: I stole this type from React Native, not sure why they use extra lines. Those aren't used but they are part of the type on mobile only. I removed them from this type, you're right that they aren't necessary and probably they create confusion.

File: shared/teams/channel/index.tsx (hunk 4 above)
Comment (jakob223) [reply shared/teams/channel/index.tsx]: Looks like bots are sorted by default in that function already.

**Comment by jakob223** [comment]
Thanks for the review @jzila ! I agree that the refactor would have been good as a separate PR 😬 .

Can you look again?

File: shared/teams/channel/index.tsx (hunk 1 above)
Comment (jzila) [reply shared/teams/channel/index.tsx]: In that case, please name it so it's only ever used in this particular spot.

File: shared/teams/channel/index.tsx (hunk 4 above)
Comment (jzila) [reply shared/teams/channel/index.tsx]: My bad, thanks for checking.

**Pull request merged**